ARM and Android fixes: version out some HardFloat tests, use TMPDIR logic and declaration from druntime, and correct comment

joakim-noah opened the pull request:
x86/x64 has no problem with unaligned reads, but the llvm optimizer gets confused by the cast in the `foreach` and assumes the `Element` is 32-bit aligned, generating ARM instructions that require word alignment and causing problems with the last unaligned test.

As for `posix_memalign`, turns out its declared in druntime but wasn't for `OSX`, so added it in dlang/druntime#1819 and used the import instead.  That druntime PR will need to be merged first.

CyberShadow commented:
As these two changes are in completely unrelated parts of the code, it would be nice to at least do them in separate commits.

ibuclaw commented:
> the llvm optimizer gets confused by the cast in the foreach and assumes the Element is 32-bit aligned,

Alternatively, you could just fix the problem with llvm. ☺️

joakim-noah commented:
The problem isn't with llvm.  The cast takes a possibly unaligned `ubyte[]` chunk and puts it in a type that is word-aligned by default in D, like `uint` or `ulong`.  That works fine on CPU arches that always allow unaligned reads, like x86, but not ARM.

If you like, try it with gdc and see if it does any better.  The last unaligned tests in the module trip up when optimized by ldc.

joakim-noah commented:
Split the changes into two commits and fixed a misnamed temp variable.

ibuclaw commented:
> That works fine on CPU arches that always allow unaligned reads, like x86, but not ARM.

Yes, I know.  It still seems strange that you are taking the `ref const`, then removing the `ref`.  Why not just make the loop `foreach (const Element; ...)`.

It's probably not just ARM, affected, so maybe some `version (NeedAlignedLoads)` might be better, that's a topic for the bikeshed anyhow.

joakim-noah commented:
>It still seems strange that you are taking the ref const, then removing the ref. Why not just make the loop foreach (const Element; ...).

I didn't want to touch anything on x86/x64, which @gchatelet did a lot of work to optimize, in #3916.

>It's probably not just ARM, affected, so maybe some version (NeedAlignedLoads) might be better, that's a topic for the bikeshed anyhow.

Sure, he mentioned MIPS also in a comment earlier in the module.  I figure it's better for those ports to add it themselves, as needed.  If you have some arch in mind, try it with gdc and let me know what needs adding.

joakim-noah commented on std/digest/murmurhash.d:
```diff
@@ -514,7 +514,15 @@ struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 6
+            {
+                // Since the block may be unaligned, this can trip up ARM
+                // codegen, so place it in a newly aligned Element first.
+                Element alignedTemp = block;
```
OK, I'm not familiar with these alignment issues and exactly how D code is likely to trigger them: I just used the simplest fix that worked.  I suppose you want [something more like @kinke's patch](https://github.com/kinke/phobos/commit/83474aeb3ee9bf089ce97c847d9d72816a881ada), which uses different types based on the alignment, though maybe he had problems with certain types that are commented out.

kinke replied:
I'm actually surprised this apparently fixed the issue for LDC, it smells like wasted optimization opportunity somewhere. A rather compact proper (I hope, I just typed it blindly) fix with no optimizations for alignments of 4/2 bytes would be:

```D
version(ARM)
{
    // data.ptr appropriately aligned for type Element?
    if ((cast(size_t) data.ptr) & (Element.alignof - 1) == 0)
    {
        foreach (ref const Element block; cast(const(Element[]))(data[0 .. remainderStart]))
            putElement(block);
    }
    else
    {
        Element alignedBlock = void;
        for (size_t i = 0; i != remainderStart; i += Element.sizeof)
        {
            (cast(ubyte*) &alignedBlock)[0 .. Element.sizeof] = data[i .. i + Element.sizeof];
            putElement(alignedBlock);
        }
    }
}
else
{
    foreach (ref const Element block; cast(const(Element[]))(data[0 .. remainderStart]))
        putElement(block);
}
```

joakim-noah commented on std/digest/murmurhash.d:
```diff
@@ -496,29 +499,76 @@ struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 6
+                    numChunks = processChunks!(ushort[size / 16])();
+                else
+                    numChunks = processChunks!(ubyte[size / 8])();
+            }
```
Lightly modified version of [@kinke's fix](https://github.com/kinke/phobos/commit/83474aeb3ee9bf089ce97c847d9d72816a881ada), let me know if this suffices.  It passes the extended tests below on ARM.

joakim-noah commented on std/digest/murmurhash.d:
```diff
@@ -496,29 +499,76 @@ struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 6
+                    numChunks = processChunks!(uint)();
+                /* TODO: cannot cast ushort/ubyte slice to Element = uint
+                else if ((startAddress & 1) == 0)
+                    numChunks = processChunks!(ushort[2])();
```
Why can't a `ushort[2]` be cast to an `int`?

kinke replied:
This still needs to be sorted out, otherwise the 32-bit hasher will still fail for unaligned data.

joakim-noah commented on std/math.d:
```diff
@@ -4590,7 +4590,7 @@ public:
 }
 
-@system unittest
+version(D_HardFloat) @system unittest
```
Note this added commit, needed when building the merge-2.074 branch of ldc.

ibuclaw replied:
Needs moving to a separate pr IMO.

joakim-noah replied:
It's so small, I'd rather just do it here.  What's the problem with this commit?  These three commits are what I need to get the tests passing on Android/ARM, fairly small PR.

ibuclaw replied:
Only one of the commits is fine, we can merge it now without waiting on the other two.  I'm not sure about this change, but I haven't looked at the body, and I don't wish to discuss here, as the topic is about murmurhash.

joakim-noah replied:
The topic is some small changes I had to make for Android/ARM, ie the contents of this PR.  The notion that one can't discuss this tiny commit here because of the surrounding discussion about murmurhash3 is laughable, as most non-negligible Phobos PRs have many such topical threads of discussion.

joakim-noah commented:
ping, made changes asked for.

kinke commented on std/digest/murmurhash.d:
```diff
@@ -739,10 +789,13 @@ version (unittest)
-        immutable unalignedHash = digest!H(data[1 .. $]); // 1 .. 1024
-        assert(alignedHash == unalignedHash);
+        immutable ubyte[1028] data = 0xAC;
+        immutable alignedHash = digest!H(data[0 .. 1023]);
```
`0 .. 1024` (contrary to the misleading previous comment) if you want to preserve the previous behavior (hashing a full 1K chunk).

joakim-noah replied:
Yeah, I thought that looked off, was just lazy to check it, corrected the last index now.  As for the alignment, it doesn't have to be but it usually is, and the right behavior will be checked regardless, as it's cycling through 5 different byte alignments.

joakim-noah commented on std/digest/murmurhash.d:
```diff
@@ -535,7 +535,14 @@ struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 6
+                        import core.stdc.string : memcpy;
+                        Element block;
+                        ()@trusted{memcpy(&block, &chunk, Element.sizeof);}();
+                    }
```
Put this in to fix the 32-bit unaligned read you pointed out is still there, @kinke, following David's suggestion.  I'm not having any issues with the way it was before or now, ie the test passes, but this won't screw up with some optimizer pass later on.

joakim-noah commented on std/digest/murmurhash.d:
```diff
@@ -496,29 +499,81 @@ struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 6
+                        import core.stdc.string : memcpy;
+                        Element block;
+                        ()@trusted{memcpy(&block, &chunk, Element.sizeof);}();
+                        putElement(block);
```
Merged and reorganized the 32-bit alignment fix I added yesterday, @kinke, here's the `memcpy` workaround David suggested earlier.  I benchmarked it on unaligned data using [@gchatelet's benchmark](https://github.com/gchatelet/murmurhash3_d/blob/master/benchmark.d), it's usually only 1-2% slower, around 4 GiB/s.

joakim-noah commented:
@ibuclaw, I'm not moving a two-line patch to a different PR, I see no reason not to discuss all three commits here.

joakim-noah commented:
Hey Iain, since you did all that work to get gdc into gcc, I'd like to point out that this PR is essentially a gift from ldc to gdc, stemming from issues that we ran into when updating to 2.072 and 2.074, as dmd won't use the two ARM-related commits anyway.

After David approved this PR, I merged the first two commits into ldc's branch of druntime, and when I tried to merge the last one into @kinke's WIP ldc branch for 2.074, I saw that he already stubbed out those tests because of other issues.

Whenever gdc updates to a newer phobos, you will likely need these patches for ARM.  If you want me to modify them in some way to better suit gdc, I'll be happy to discuss that.  Until then, this PR is only here to benefit gdc, as ldc is already using it.

ZombineDev commented:
(Closing & reopening the PR in order to restart the CI tests.)

ZombineDev commented on std/digest/murmurhash.d:
```diff
@@ -739,10 +794,13 @@ version (unittest)
-        assert(alignedHash == unalignedHash);
+        immutable ubyte[1028] data = 0xAC;
+        immutable alignedHash = digest!H(data[0 .. 1024]);
+        foreach(i; 1 .. 5)
```
CirclCI fails because there needs to be a space between `foreach` and `(`.

ZombineDev commented:
> Sizeof BCValue: 40LU
60 opcodes remaining
/dev/shm/dtest/work/repo/dlang.org/../dmd/src/ddmd/ctfe/bc_skeleton.d(9): Error: undefined identifier `BCFunction`
SDCs LLVM Header are not avilable
LLVM_Backend is not compiled
BCGen
immutable(const(BCValue) function(BCValue[], BCHeap*) @safe)
C_BCGen
immutable(BCValue function(BCValue[], BCHeap*) pure @safe)
posix.mak:527: recipe for target '.generated/docs-prerelease.json' failed
make: *** [.generated/docs-prerelease.json] Error

@CyberShadow it looks like DAutoTest is trying to build the newCTFE dmd branch, instead of master or stable. Any ideas? Probably restarting the test would fix the error. Is there a way I can do that without force-pushing in Joakim's branch? (I'm on the phone currently and I don't have a proper build environment setup).

joakim-noah commented:
Thanks for the spacing tip, Petar, added it to my commit and removed yours.

joakim-noah commented on std/file.d:
```diff
@@ -4293,11 +4293,6 @@ string tempDir() @trusted
-        {
-            // Don't check for a global temporary directory as
-            // Android doesn't have one.
-        }
```
Added another Android commit, as ldc is now packaged on Android itself, and the Termux app does set `TMPDIR`.  This was versioned out because it was assumed that most would run D code from their own Android app, ie an apk, which don't have any global temp directory.  However, now that D can be used from the command-line on Android itself, with an app that sets `TMPDIR` for all command-line D tools, just use the below Posix logic.  You will have to set [the tempDir by hand for each apk anyway](https://gist.github.com/joakim-noah/09a59bdd6e04dd2f3d3ad6e15573ad97#file-phobos_1-3_ldc_arm-L9), if you use it. 

This already hit me when building [`rdmd` for Android, where I had to patch it to check `TMPDIR`](https://github.com/termux/termux-packages/pull/1377/files#diff-267b88bd99296808606d2a526c27d831), can take that out with this change.

joakim-noah commented:
Removed the largest murmurhash commit, as it was originally written by @kinke and he has since reworked it and submitted a new version in another pull. Also added one more commit with a doc fix, this pull is very small and easy to review now, @wilzbach.
